src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -250,122 +224,37 @@ private void saveModelCheckpointAsync(Map<String, Object> source, String modelId
-
-    private void flush(BulkRequest bulkRequest) {
-        clientUtil.<BulkRequest, BulkResponse>execute(BulkAction.INSTANCE, bulkRequest, ActionListener.wrap(r -> {
-            if (r.hasFailures()) {
-                requests.addAll(BulkUtil.getIndexRequestToRetry(bulkRequest, r));
-            } else if (requests.size() >= maxBulkRequestSize / 2) {
-                // during maintenance, we may have much more waiting in the queue.
-                // trigger another flush if that's the case.
-                flush();
-            }
-        }, e -> {
-            logger.error("Failed bulking checkpoints", e);
-            // retry during next bulk.
-            for (DocWriteRequest<?> req : bulkRequest.requests()) {
-                requests.add(req);
-            }
-        }));
-    }
-
-    /**
-     * Prepare bulking the input model state to the checkpoint index.
-     * We don't save checkpoints within checkpointInterval again.
-     * @param modelState Model state
-     * @param modelId Model Id
-     */
-    public void write(ModelState<EntityModel> modelState, String modelId) {
-        write(modelState, modelId, false);
-    }
-
-    /**
-     * Prepare bulking the input model state to the checkpoint index.
-     * We don't save checkpoints within checkpointInterval again, except this
-     * is from cold start. This method will update the input state's last
-     *  checkpoint time if the checkpoint is staged (ready to be written in the
-     *  next batch).
-     * @param modelState Model state
-     * @param modelId Model Id
-     * @param coldStart whether the checkpoint comes from cold start
+     * Prepare for index request using the contents of the given model state
+     * @param modelState an entity model state
+     * @return serialized JSON map or empty map if the state is too bloated
+     * @throws IOException  when serialization fails
      */
-    public void write(ModelState<EntityModel> modelState, String modelId, boolean coldStart) {
-        Instant instant = modelState.getLastCheckpointTime();
-        // Instant.MIN is the default value. We don't save until we are sure.
-        if ((instant == Instant.MIN || instant.plus(checkpointInterval).isAfter(clock.instant())) && !coldStart) {
-            return;
+    public Map<String, Object> toIndexSource(ModelState<EntityModel> modelState) throws IOException {
+        Map<String, Object> source = new HashMap<>();
+        EntityModel model = modelState.getModel();
+        String serializedModel = toCheckpoint(model);
+        if (serializedModel == null || serializedModel.length() > maxCheckpointBytes) {
+            logger
+                .warn(
+                    new ParameterizedMessage(
+                        "[{}]'s model empty or too large: [{}] bytes",
+                        modelState.getModelId(),
+                        serializedModel == null ? 0 : serializedModel.length()
+                    )
+                );
+            return source;
```

Comment:
Why return an empty map instead of throwing an exception?

Reply:
that can be an option.  It is two styles: https://docs.python.org/3/glossary.html#term-lbyl vs https://docs.python.org/3/glossary.html#term-eafp

For me, I am trying to avoid exceptions since it is more expensive than a conditional jump.

Reply:
Ah I see that makes sense.

Why then return an empty map instead of null?

Reply:
Best practices is not to return null.  Return an empty collection or Optional is recommended by the book Effective Java.